New thoughts and thought constraints

Snowstar38 opened the pull request:
<!-- Write BELOW The Headers and ABOVE The comments else it may not be viewable. -->
<!-- You can view CONTRIBUTING.md for a detailed description of the pull request process. -->
<!-- Be sure to name your PR something descriptive and succinct; include Bugfix: Feature: Enhancement: or Content: in the title to describe what type of PR it is. -->

## About The Pull Request

Added three new thought constraints for more thought variety:

- **clan_missing_status_constraint**: for detecting if the Clan is missing a leader/deputy/medcat

This is because the random_cat method for thoughts can only check if the random cat chosen has a given status. It can't actually detect if a status is absent from the Clan altogether.

- **herb_max_constraint**: For thoughts related to herb stores. I do plan to find a way to calculate the needed herb amounts for the Clan's population in the future.

- **leader_life_constraint**: For thoughts related to how many lives the leader has. I made it so it can take multiple values as a way to avoid needing both a max and a min constraint; let me know if you think it should be done differently!

## Why This Is Good For ClanGen

It's always nice to have more thoughts directly related to the current state of the Clan. These will make the cats feel more connected to what's happening in the game.

## Proof of Testing

![image](https://github.com/user-attachments/assets/51fff8ae-82bb-486f-a7d7-0f460a579514)
![image](https://github.com/user-attachments/assets/c086544c-dfc2-4550-a671-63982502373a)
![image](https://github.com/user-attachments/assets/90b61965-bbd2-463b-84a5-d8588f09598e)


<!-- Include any screenshots, debugging steps, or links to beta testing threads here. At least one form of proof of testing is REQUIRED for all new content. You must be able to run the code locally before you PR it here. -->

## Changelog/Credits

Added new thoughts:

- Thoughts for when the Clan has no leader, deputy, and/or medicine cat
- Thoughts about the herb stores
- Thoughts about the leader's remaining lives

Snowstar38 committed:
New thought constraints

- clan_missing_status_constraint: for detecting if the Clan is missing a leader/deputy/medcat

This is because the random_cat method for thoughts can only check if the random cat chosen has a given status. It can't actually detect if a status is absent altogether.

- herb_max_constraint: For thoughts related to herb stores. I do plan to find a way to calculate the needed herb amounts for the Clan's population one day, but right now, it can still detect if the store is empty or not.

- leader_life_constraint: For thoughts related to how many lives the leader has. I made it so it can take multiple values as a way to avoid needing both a max and a min constraint, let me know if you think it should be done differently!

Snowstar38 committed:
Bug fixing

Snowstar38 committed:
Update medicine_cat.json

Snowstar38 committed:
cleanup

scribblecrumb commented on scripts/cat/thoughts.py:
```diff
@@ -94,7 +95,36 @@ def cats_fulfill_thought_constraints(main_cat, random_cat, thought, game_mode, b
+        
+        # Constraints for empty positions of power in the Clan
+        # Only works for leader, deputy, and medicine cat
+        if 'clan_missing_status_constraint' in thought:
```
I'd like to see this be applicable to all ranks and I believe it would be relatively simple.  The PatrolEvent and ShortEvent formats already utilize a similar constraint system, however theirs is done via tags and then is filtered via the `get_alive_status_cats()` function in `utility.py`.  Plugging the list from `clan_missing_status_constraint` into `get_alive_status_cats()` would work well, since the function already takes a list of statuses as a kwarg.

It might additionally be wise to create a `clan_present_status_constraint` to handle the opposite situations: where you want to check that a certain status *is* present in the clan.

Snowstar38 replied:
Wow, thank you! I had no idea that utility function existed. Now it works much more smoothly!

scribblecrumb commented on scripts/cat/thoughts.py:
```diff
@@ -94,7 +95,36 @@ def cats_fulfill_thought_constraints(main_cat, random_cat, thought, game_mode, b
+        if 'herb_max_constraint' in thought:
+            if game.clan:
+                herb_max = (thought['herb_max_constraint'])[0]
+                if herb_max < sum(game.clan.herbs.values()):
```
Regarding calculating how many herbs a clan needs, I'd direct you toward the calculations used in ShortEvent filtering.  ShortEvents allow constraining to set supply amounts of both certain herbs and overall herb supply: low, adequate, full, excess, as well as an "always" to trigger no matter how many herbs are in the supply.  I'll admit the filtering is a bit of a headache to look at, so if you want some help understanding it then let me know.

This is admittedly something I'm still trying to balance sufficiently.  If we're going to have it determined in multiple places, then there should be consideration towards either a function or some game_config settings so that balancing can remain consistent and easy to alter in the future.

Snowstar38 replied:
I tried adding some utility functions based off of the values in the medicine cat concern string (where it shows what they think of their current stores), is that more like what you had in mind?

scribblecrumb replied:
Yeah, very nice!  Could you also allow for checking the supply of a singular herb?

scribblecrumb commented on resources/dicts/thoughts/alive/warrior.json:
```diff
@@ -34,12 +34,36 @@
+    {
+        "id": "warrior_gen_no_deputy_ambitious",
+        "thoughts": [
+            "Desperately hopes r_c will choose them to be the new deputy"
```
```suggestion
            "Desperately hopes r_c will choose {PRONOUN/m_c/object} to be the new deputy"
```

scribblecrumb commented on resources/dicts/thoughts/alive/leader.json:
```diff
@@ -296,5 +296,38 @@
+        "id": "leader_no_deputy",
+        "thoughts": [
+            "Is carefully mulling over who would be a good deputy",
+            "Organizes the patrols by {PRONOUN/m_c/object}"
```
```suggestion
            "Organizes the patrols by {PRONOUN/m_c/self}"
```

scribblecrumb commented on resources/dicts/thoughts/alive/general.json:
```diff
@@ -4607,5 +4607,89 @@
+    {
+        "id": "skilled_no_medcat_gen",
+          "thoughts": [
+              "Wonders if they should become a medicine cat"
```
```suggestion
              "Wonders if {PRONOUN/m_c/subject} should become a medicine cat"
```

scribblecrumb reviewed (changes requested):
Mostly feedback looking to make your additional constraints more similar to their event counterparts, for the sake of consistency.

Snowstar38 committed:
Update resources/dicts/thoughts/alive/warrior.json

Co-authored-by: scribblecrumb <115410010+scribblecrumb@users.noreply.github.com>

Snowstar38 committed:
Update resources/dicts/thoughts/alive/general.json

Co-authored-by: scribblecrumb <115410010+scribblecrumb@users.noreply.github.com>

Snowstar38 committed:
Update resources/dicts/thoughts/alive/leader.json

Co-authored-by: scribblecrumb <115410010+scribblecrumb@users.noreply.github.com>

Snowstar38 committed:
get_alive_status_cats to the rescue

Snowstar38 committed:
Update warrior.json

Snowstar38 committed:
Add get_herb_supply_level and check_herb_constraint

Snowstar38 committed:
Update medicine_cat.json

Snowstar38 committed:
Update warrior.json

Snowstar38 committed:
Update kitten.json

Snowstar38 committed:
Update general.json

Snowstar38 committed:
Update general.json

Snowstar38 committed:
Fix for first load-in

Since game.clan doesn't exist when generating thoughts for the first load-in, the checks need to return False

Snowstar38 committed:
remove test prints

Snowstar38 committed:
Update thoughts.py

j-gynn commented:
Closing as stale.
